rtos/ConditionVariable.h

```diff
@@ -35,26 +35,44 @@ namespace rtos {
 
 struct Waiter;
 
-/** This class provides a safe way to wait for or send notifications of condition changes
+/** The ConditionVariable class is a synchronization primitive that allows
+ *   threads to wait until a particular condition occurs.
  *
- * This class is used in conjunction with a mutex to safely wait for or
- * notify waiters of condition changes to a resource accessible by multiple
+ * The condition variable is used in conjunction with a mutex to safely wait for
+ * or notify waiters of condition changes to a resource accessible by multiple
  * threads.
  *
- * # Defined behavior
- * - All threads waiting on the condition variable wake when
+ * The thread that intends to wait on a ConditionVariable must:
+ * - Acquire a lock on a mutex
+ * - Execute `wait`, `wait_for`, or `wait_until`. While the thread is waiting,
+ *   the mutex will be unlocked.
+ * - When the condition variable has been notified, or in the case of `wait_for`
+ *   and `wait_until` the timeout expires the thread is awakened.
+ *
+ * The thread that intends to notify a ConditionVariable must:
+ * - Acquire a lock on the mutex used to construct the condition variable.
+ * - Execute `notify_one` or `notify_all` on the condition variable.
+ *
+ * ## Defined behavior
+ * - All threads that are waiting on the condition variable will wake when
  *   ConditionVariable::notify_all is called.
- * - If one or more threads are waiting on the condition variable at least
- *   one of them wakes when ConditionVariable::notify is called.
+ * - At least one thread that is waiting on the condition variable will wake
+ *   when ConditionVariable::notify_one is called.
+ * - While waiting for a thread is waiting for notification of a
+ *   ConditionVariable it will release the lock held on the mutex.
+ * - The ConditionVariable will reacquire the mutex lock before exiting the wait
+ *   function.
  *
- * # Undefined behavior
+ * ## Undefined behavior
```

Comment:
I wonder if there should be a split between "undefined" and "unspecified" behaviour here.

Undefined behaviour covers "nasal demons" in standardese - never do this cos *anything* could happen. Most of this stuff is just "unspecified" - eg one thread will definitely wake, but we don't specify which.

Will annotate which ones would need to be changed to unspecified